### PR TITLE
Enable React concurrent mode

### PR DESCRIPTION
**Pull request by Kicu**
### React concurrent mode
Starting from react 18 we can run React in new concurrent way (which seems the way React team is moving). Concurrent mode enables a lot of new behaviours in react, most importantly renders can be interrupted by React, re-run or run more than once. This is supposed to make react more performant and webapps more responsive to user actions.
State of this in Expensify:
 - on native apps after we migrated new arch and enabled Fabric - to my knowledge - we should already be running react concurrently
 - on web we are running legacy mode and **this PR** enables concurrent mode

Further reading: 
 - https://vercel.com/blog/how-react-18-improves-application-performance and also
 - https://react.dev/blog/2022/03/29/react-v18#what-is-concurrent-react

### React StrictMode
`<StrictMode>` is a component that wraps the whole App in (or parts of App) and it runs extra checks and extra behaviors **only in dev**. So in essence it's a kind of devtool for developers, that should help catch tricky bugs that _could_ be introduced when running concurrent mode. Mainly it renders every component twice (discarding the first render) and runs all `effects` twice.
This behaviour ☝️ of StrictMode is the source of most bugs that happen to us.

Note: `StrictMode` is **not needed** for concurrent react to work but it is much safer to use it. 

https://react.dev/reference/react/StrictMode

## Details
This PR enables running React 18+ in concurrent mode alongside using StrictMode.
I did not provide screenshots or videos because this affects the whole how React behaves, so in theory affects the whole app.

### Done
 - react concurrent mode turned on and whole App wrapped in `<StrictMode>`
 - react-navigation fixed via patch; see details here: https://github.com/Expensify/App/pull/42592#issuecomment-2178832015
 - `react-native-reanimated` fixed via patch, however it should be released soon as the team is aware of StrictMode problems.
 The related upstream PRs are: https://github.com/software-mansion/react-native-reanimated/pull/6129 and https://github.com/software-mansion/react-native-reanimated/pull/6123 and reanimated team expects correct version to be released in 2.13 

#### List of known bugs & fixes
 - ~`react-navigation` breaks on `StrictMode` - fixed via local patch - need to verify the legitness of said patch~ solved
 - `react-native-reanimated`
   - animations not working - OS team will fix, we have local patch in the meantime
   - ~initial animation when entering Report sometimes flickers; this happens because of StrictMode so only in dev~

### Fixed Issues

$ https://github.com/Expensify/App/issues/33531
PROPOSAL:

### Tests

- Verify that web and native build and run
- test all the most important features of the app and verify nothing breaks

### QA Steps

 - verify that web and native app runs
 - test all the most important features of the app and verify that nothing obvious breaks

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: src/components/TextInput/TextInputClearButton/index.tsx
Comment (Kicu) [inline src/components/TextInput/TextInputClearButton/index.tsx:43]: this is unrelated but a source of warning in the browser console. While working on updated React I wanted to get rid of as many console warnings as I can since it helps to find the actual React bugs

**Comment by Kicu** [comment]
### `react-navigation` status
 - react-navigation in version 6 (current stable) does not work well in `StrictMode` when using linkingConfig - which we do
 - I created a simple RN app with reproduction of this behaviour and I have tested it using newer v7 version and it works, However react-navigation 7 is still in alpha: https://github.com/react-navigation/react-navigation/releases
 - I asked @satya164 who is an author of react navigation and he said that it is unlikely for this fix to be backported to 6 and suggested we try version 7 - perhaps this is a good idea to try
 - I have found the source of bug in v6 and compared the code to v7 and I was able to prepare a patch for expensify. I can't 100% guarantee at this point that the patch won't break something else, but it feels mostly safe - I only edited one file.

### detailed description of patch
The source of problems for us is that inside `useNavigationBuilder` there is code that cleans state on component unmount here: https://github.com/react-navigation/react-navigation/blob/6.x/packages/core/src/useNavigationBuilder.tsx#L557-L561
Because of `StrictMode` navigation components will be rendered twice and their effects will run twice. This in turn means that when the navigation renders and runs for the second time, the state is already cleaned from the previous effect cleanup.

Then this check will trigger: https://github.com/react-navigation/react-navigation/blob/6.x/packages/core/src/useNavigationBuilder.tsx#L372 and there is no navigation happening at all.

The patch fixes this in a similar way to v7 branch.

CC @adamgrzybowski @WoLewicki for visibility

**Comment by adamgrzybowski** [comment]
As a follow up to @Kicu mentioning react-navigation 7. This version introduces two functionalities that we may be interested in:
- preload -> We could potentially use it for optimization and performance.
- popTo -> It may help us with simplifying the UP action implementation.

**Comment by Kicu** [comment]
@mountiny @roryabraham We're ready for liftoff 🚀 

Web and native builds work, there are no crucial issues and app works. Now I think we need some heavy testing by QA's - I tested only the basic functionalities that I know about. If any bugs come up they could be rather tricky or non-obvious.

**Comment by WoLewicki** [comment]
Maybe we could open another PR with all the changes not strictly related to the `StrictMode` so they can be also tested and merged earlier if they work properly? This way, this PR would focus only on the `StrictMode`. And that other PR is merged before this one, you can just merge the newest `main` here. Wdyt @Kicu ?

**Comment by Kicu** [comment]
I'm not sure. Technically it will be easy to split this into `only concurrent mode` and `only StrictMode` and I could do it, but how much would this really improve the process?
In theory if QA people will test this on production build, then strictmode should not affect it in any way. Per docs: 
> All of these checks are development-only and do not impact the production build.

So maybe leaving this PR as is will just be simpler?

**Comment by WoLewicki** [comment]
I mean we should leave all the changes in this PR for sure. I'm talking changes such as this: https://github.com/Expensify/App/pull/42592/files#diff-536cf8af84bdec51c0d64f73e52c8f60523c348a0a1d229af3882de2b8bdf5ffR43 to be also introduced in other PR that can be quickly merged so this one is cleaner.

**Comment by mountiny** [comment]
Running a build https://github.com/Expensify/App/actions/runs/9597072039

@rayane-djouah are you able to do testing on this one and report any bugs you find? thanks!

**Comment by mountiny** [comment]
This is how the android failed https://github.com/Expensify/App/actions/runs/9597286693/job/26466130194 Does not seem related to the changes here

**Comment by WoLewicki** [comment]
Yeah, would be best to get another build. Also, the PR should ideally only introduce changes on `web` after @Kicu makes another PR with "global" changes.

**Comment by Kicu** [comment]
@WoLewicki @mountiny what we discussed is done here: https://github.com/Expensify/App/pull/44155

**Review by roryabraham — CHANGES_REQUESTED** [review]
For both of the added patches, please link clearly in the PR description to the upstream PR for the patch.

File: src/components/TextInput/TextInputClearButton/index.tsx
Comment (roryabraham) [reply src/components/TextInput/TextInputClearButton/index.tsx:43]: Sounds good, but may I suggest creating a separate PR to remove this ref?

File: src/setup/platformSetup/index.website.ts
Comment (blazejkustra) [inline src/setup/platformSetup/index.website.ts:3]: Why was this removed here?

**Review by rayane-d — CHANGES_REQUESTED** [review]

File: src/setup/platformSetup/index.website.ts
Comment (Kicu) [reply src/setup/platformSetup/index.website.ts:3]: will do 👍

**Comment by Kicu** [comment]
@rayane-djouah PR cleaned up

**Comment by Kicu** [comment]
@roryabraham 
Updated PR description with `reanimated` PRs that should soon be merged.

As for `react-navigation` patch the situation is a bit different.
The bug for StrictMode is in version 6 but in version 7 this bug is fixed and the lib works.

Citing myself from here: https://github.com/Expensify/App/pull/42592#issuecomment-2178832015 
> I asked @satya164 who is an author of react navigation and he said that it is unlikely for this fix to be backported to 6 and suggested we try version 7 - perhaps this is a good idea to try

Fixed on v7, however 7 is still in alpha :/
I can try to open the PR into 6.x if you want me to, but not sure how likely it is that any PR to version 6 will be merged. Right now I have tested the patch with expensify codebase - I have no idea whether its a universally correct solution for react-navigation or not.
What do you want me to do next?

**Comment by Kicu** [comment]
All the changes in this PR that were separate got merged in here: https://github.com/Expensify/App/pull/44155

This PR is now purely adding react concurrent mode and `StrictMode` to App, with the patches that make it work

**Review by blazejkustra — APPROVED** [review]
Awesome work @Kicu!

**Review by rayane-d — APPROVED** [review]
Thank you!
Changes look good to me and test well. The reviewer checklist is also complete.
:ribbon::eyes::ribbon: C+ reviewed

**Comment by roryabraham** [comment]
> The bug for StrictMode is in version 6 but in version 7 this bug is fixed and the lib works.

This is great context, it means that the patch can be removed if/when we upgrade to React Navigation v7+

**Review by roryabraham — APPROVED** [review]

**Comment by jasperhuangg** [comment]
@Kicu @roryabraham I'm having trouble following how this code only runs on dev considering that the [`USE_REACT_STRICT-MODE` flag](https://github.com/Expensify/App/blob/390606a0336a880ede5de3fb9f38729afa5f060e/src/CONFIG.ts#L99) is always `true` and isn't dependent on any environment variables. Can you explain?

I'm mainly asking because there's a crash happening [here](https://github.com/Expensify/App/issues/44583) that involves a component [exceeding the max update depth](https://react.dev/errors/185?invariant=185).

**Comment by jasperhuangg** [comment]
Hmm taking a deeper look at the docs it seems there's logic within React's StrictMode implementation that prevents it from running effects twice on staging/production builds.

**Comment by Kicu** [comment]
@jasperhuangg indeed its the internal behavior of StrictMode, described in the React docs that makes it only work in dev.
https://react.dev/reference/react/StrictMode#enabling-strict-mode-for-entire-app
So I'm trusting react devs here. It should be safe to ship apps wrapped with strict mode to prod and expect strict mode will not do any of the checks.

The CONFIG flag is for the convenience of developers for quick testing locally why their code fails

**Comment by ikevin127** [comment]
👋  The introduction of `StrictMode` with this PR created the following **dev** issue:
- https://github.com/Expensify/App/issues/44566

The issue was fixed by PR https://github.com/Expensify/App/pull/45559, more details on the proposal [here](https://github.com/Expensify/App/issues/44566#issuecomment-2199623262).

**Pull request merged**
